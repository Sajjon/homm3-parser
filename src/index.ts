import { parse, StreamInput, TagOrWrapper } from 'binary-markup'
import * as pako from 'pako'
import { h3mFile } from './parsers/h3m/h3m.bml'
import { H3MFile } from './parsers/h3m/contracts'

export * from './parsers/assets'
export * from './parsers/h3m/contracts'
export * from './parsers/h3m/constants'

export function parseH3M(data: StreamInput): H3MFile {
  return parse<H3MFile>(h3mFile, data)
}

export function parseH3MFile(data: Uint8Array): H3MFile {
  const content = pako.inflate(data)
  return parseH3M(content.buffer)
}

export function parseAsset<T>(type: TagOrWrapper<T>, data: StreamInput): T {
  return parse(type, data)
}
