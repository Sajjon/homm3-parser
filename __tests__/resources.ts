import { testMap } from './utils'
import { H3MFile } from '../src'

describe('resources', () => {
  testMap('resources', 'resources.h3m', (map: H3MFile) => map.objectDetails)
})
