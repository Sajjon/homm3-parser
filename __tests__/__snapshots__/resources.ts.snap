// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`resources should parse resources 1`] = `
Object {
  "count": 24,
  "entries": Array [
    Object {
      "attributes": Object {
        "above": 0,
        "active": Array [
          0,
          0,
          0,
          0,
          0,
          128,
        ],
        "allowedLandscapes": 511,
        "header": "AVTchst0.def",
        "landscapeGroup": 255,
        "metaType": "Treasure",
        "objectGroup": "Treasure",
        "objectSubType": 0,
        "objectType": "TreasureChest",
        "passable": Array [
          255,
          255,
          255,
          255,
          255,
          127,
        ],
      },
      "body": Object {
        "data": null,
        "type": "Treasure",
      },
      "oaIndex": 2,
      "x": 1,
      "y": 1,
      "z": 0,
    },
    Object {
      "attributes": Object {
        "above": 0,
        "active": Array [
          0,
          0,
          0,
          0,
          0,
          128,
        ],
        "allowedLandscapes": 511,
        "header": "AVTcrys0.def",
        "landscapeGroup": 1,
        "metaType": "Resource",
        "objectGroup": "Treasure",
        "objectSubType": 4,
        "objectType": "Resource",
        "passable": Array [
          255,
          255,
          255,
          255,
          255,
          127,
        ],
      },
      "body": Object {
        "data": Object {
          "hasGuardians": Object {
            "creatures": null,
            "hasCreatures": false,
            "message": "crystal",
          },
          "quantity": 0,
        },
        "type": "Resource",
      },
      "oaIndex": 3,
      "x": 3,
      "y": 1,
      "z": 0,
    },
    Object {
      "attributes": Object {
        "above": 0,
        "active": Array [
          0,
          0,
          0,
          0,
          0,
          128,
        ],
        "allowedLandscapes": 511,
        "header": "AVTgems0.def",
        "landscapeGroup": 255,
        "metaType": "Resource",
        "objectGroup": "Treasure",
        "objectSubType": 5,
        "objectType": "Resource",
        "passable": Array [
          255,
          255,
          255,
          255,
          255,
          127,
        ],
      },
      "body": Object {
        "data": Object {
          "hasGuardians": Object {
            "creatures": null,
            "hasCreatures": false,
            "message": "Gems",
          },
          "quantity": 0,
        },
        "type": "Resource",
      },
      "oaIndex": 4,
      "x": 5,
      "y": 1,
      "z": 0,
    },
    Object {
      "attributes": Object {
        "above": 0,
        "active": Array [
          0,
          0,
          0,
          0,
          0,
          128,
        ],
        "allowedLandscapes": 511,
        "header": "AVTgold0.def",
        "landscapeGroup": 1,
        "metaType": "Resource",
        "objectGroup": "Treasure",
        "objectSubType": 6,
        "objectType": "Resource",
        "passable": Array [
          255,
          255,
          255,
          255,
          255,
          127,
        ],
      },
      "body": Object {
        "data": Object {
          "hasGuardians": Object {
            "creatures": null,
            "hasCreatures": false,
            "message": "gold",
          },
          "quantity": 0,
        },
        "type": "Resource",
      },
      "oaIndex": 5,
      "x": 7,
      "y": 1,
      "z": 0,
    },
    Object {
      "attributes": Object {
        "above": 0,
        "active": Array [
          0,
          0,
          0,
          0,
          0,
          128,
        ],
        "allowedLandscapes": 511,
        "header": "AVTmerc0.def",
        "landscapeGroup": 1,
        "metaType": "Resource",
        "objectGroup": "Treasure",
        "objectSubType": 1,
        "objectType": "Resource",
        "passable": Array [
          255,
          255,
          255,
          255,
          255,
          127,
        ],
      },
      "body": Object {
        "data": Object {
          "hasGuardians": Object {
            "creatures": null,
            "hasCreatures": false,
            "message": "mercury",
          },
          "quantity": 0,
        },
        "type": "Resource",
      },
      "oaIndex": 6,
      "x": 9,
      "y": 1,
      "z": 0,
    },
    Object {
      "attributes": Object {
        "above": 0,
        "active": Array [
          0,
          0,
          0,
          0,
          0,
          128,
        ],
        "allowedLandscapes": 511,
        "header": "AVTore0.def",
        "landscapeGroup": 1,
        "metaType": "Resource",
        "objectGroup": "Treasure",
        "objectSubType": 2,
        "objectType": "Resource",
        "passable": Array [
          255,
          255,
          255,
          255,
          255,
          127,
        ],
      },
      "body": Object {
        "data": Object {
          "hasGuardians": Object {
            "creatures": null,
            "hasCreatures": false,
            "message": "ore",
          },
          "quantity": 0,
        },
        "type": "Resource",
      },
      "oaIndex": 7,
      "x": 11,
      "y": 1,
      "z": 0,
    },
    Object {
      "attributes": Object {
        "above": 0,
        "active": Array [
          0,
          0,
          0,
          0,
          0,
          128,
        ],
        "allowedLandscapes": 511,
        "header": "AVTsulf0.def",
        "landscapeGroup": 1,
        "metaType": "Resource",
        "objectGroup": "Treasure",
        "objectSubType": 3,
        "objectType": "Resource",
        "passable": Array [
          255,
          255,
          255,
          255,
          255,
          127,
        ],
      },
      "body": Object {
        "data": Object {
          "hasGuardians": Object {
            "creatures": null,
            "hasCreatures": false,
            "message": "sulfur",
          },
          "quantity": 0,
        },
        "type": "Resource",
      },
      "oaIndex": 8,
      "x": 13,
      "y": 1,
      "z": 0,
    },
    Object {
      "attributes": Object {
        "above": 0,
        "active": Array [
          0,
          0,
          0,
          0,
          0,
          128,
        ],
        "allowedLandscapes": 511,
        "header": "AVTwood0.def",
        "landscapeGroup": 1,
        "metaType": "Resource",
        "objectGroup": "Treasure",
        "objectSubType": 0,
        "objectType": "Resource",
        "passable": Array [
          255,
          255,
          255,
          255,
          255,
          127,
        ],
      },
      "body": Object {
        "data": Object {
          "hasGuardians": Object {
            "creatures": null,
            "hasCreatures": false,
            "message": "wood",
          },
          "quantity": 0,
        },
        "type": "Resource",
      },
      "oaIndex": 9,
      "x": 1,
      "y": 3,
      "z": 0,
    },
    Object {
      "attributes": Object {
        "above": 0,
        "active": Array [
          0,
          0,
          0,
          0,
          0,
          128,
        ],
        "allowedLandscapes": 511,
        "header": "AVTrndm0.def",
        "landscapeGroup": 1,
        "metaType": "Resource",
        "objectGroup": "Treasure",
        "objectSubType": 0,
        "objectType": "RandomResource",
        "passable": Array [
          255,
          255,
          255,
          255,
          255,
          127,
        ],
      },
      "body": Object {
        "data": Object {
          "hasGuardians": null,
          "quantity": 0,
        },
        "type": "Resource",
      },
      "oaIndex": 10,
      "x": 3,
      "y": 3,
      "z": 0,
    },
    Object {
      "attributes": Object {
        "above": 0,
        "active": Array [
          0,
          0,
          0,
          0,
          0,
          128,
        ],
        "allowedLandscapes": 511,
        "header": "AVTrndm0.def",
        "landscapeGroup": 1,
        "metaType": "Resource",
        "objectGroup": "Treasure",
        "objectSubType": 0,
        "objectType": "RandomResource",
        "passable": Array [
          255,
          255,
          255,
          255,
          255,
          127,
        ],
      },
      "body": Object {
        "data": Object {
          "hasGuardians": Object {
            "creatures": Array [
              Object {
                "quantity": 1,
                "type": "Pikeman",
              },
              Object {
                "quantity": 0,
                "type": "EMPTY",
              },
              Object {
                "quantity": 0,
                "type": "EMPTY",
              },
              Object {
                "quantity": 0,
                "type": "EMPTY",
              },
              Object {
                "quantity": 0,
                "type": "EMPTY",
              },
              Object {
                "quantity": 0,
                "type": "EMPTY",
              },
              Object {
                "quantity": 2,
                "type": "Griffin",
              },
            ],
            "hasCreatures": true,
            "message": "with guardians",
          },
          "quantity": 10,
        },
        "type": "Resource",
      },
      "oaIndex": 10,
      "x": 1,
      "y": 6,
      "z": 0,
    },
    Object {
      "attributes": Object {
        "above": 0,
        "active": Array [
          0,
          0,
          0,
          0,
          0,
          64,
        ],
        "allowedLandscapes": 511,
        "header": "AVXabnd0.def",
        "landscapeGroup": 1,
        "metaType": "ResourceGenerator",
        "objectGroup": "",
        "objectSubType": 7,
        "objectType": "Mine",
        "passable": Array [
          255,
          255,
          255,
          255,
          255,
          31,
        ],
      },
      "body": Object {
        "data": Object {
          "owner": 2,
        },
        "type": "ResourceGenerator",
      },
      "oaIndex": 11,
      "x": 3,
      "y": 14,
      "z": 0,
    },
    Object {
      "attributes": Object {
        "above": 0,
        "active": Array [
          0,
          0,
          0,
          0,
          0,
          64,
        ],
        "allowedLandscapes": 511,
        "header": "AVXabnd0.def",
        "landscapeGroup": 1,
        "metaType": "ResourceGenerator",
        "objectGroup": "",
        "objectSubType": 7,
        "objectType": "Mine",
        "passable": Array [
          255,
          255,
          255,
          255,
          255,
          31,
        ],
      },
      "body": Object {
        "data": Object {
          "owner": 4,
        },
        "type": "ResourceGenerator",
      },
      "oaIndex": 11,
      "x": 7,
      "y": 14,
      "z": 0,
    },
    Object {
      "attributes": Object {
        "above": 0,
        "active": Array [
          0,
          0,
          0,
          0,
          0,
          64,
        ],
        "allowedLandscapes": 511,
        "header": "AVXabnd0.def",
        "landscapeGroup": 1,
        "metaType": "ResourceGenerator",
        "objectGroup": "",
        "objectSubType": 7,
        "objectType": "Mine",
        "passable": Array [
          255,
          255,
          255,
          255,
          255,
          31,
        ],
      },
      "body": Object {
        "data": Object {
          "owner": 72,
        },
        "type": "ResourceGenerator",
      },
      "oaIndex": 11,
      "x": 11,
      "y": 14,
      "z": 0,
    },
    Object {
      "attributes": Object {
        "above": 0,
        "active": Array [
          0,
          0,
          0,
          0,
          0,
          64,
        ],
        "allowedLandscapes": 511,
        "header": "AVXabnd0.def",
        "landscapeGroup": 1,
        "metaType": "ResourceGenerator",
        "objectGroup": "",
        "objectSubType": 7,
        "objectType": "Mine",
        "passable": Array [
          255,
          255,
          255,
          255,
          255,
          31,
        ],
      },
      "body": Object {
        "data": Object {
          "owner": 16,
        },
        "type": "ResourceGenerator",
      },
      "oaIndex": 11,
      "x": 15,
      "y": 14,
      "z": 0,
    },
    Object {
      "attributes": Object {
        "above": 0,
        "active": Array [
          0,
          0,
          0,
          0,
          0,
          64,
        ],
        "allowedLandscapes": 511,
        "header": "AVXabnd0.def",
        "landscapeGroup": 1,
        "metaType": "ResourceGenerator",
        "objectGroup": "",
        "objectSubType": 7,
        "objectType": "Mine",
        "passable": Array [
          255,
          255,
          255,
          255,
          255,
          31,
        ],
      },
      "body": Object {
        "data": Object {
          "owner": 32,
        },
        "type": "ResourceGenerator",
      },
      "oaIndex": 11,
      "x": 19,
      "y": 14,
      "z": 0,
    },
    Object {
      "attributes": Object {
        "above": 0,
        "active": Array [
          0,
          0,
          0,
          0,
          0,
          64,
        ],
        "allowedLandscapes": 511,
        "header": "AVXabnd0.def",
        "landscapeGroup": 1,
        "metaType": "ResourceGenerator",
        "objectGroup": "",
        "objectSubType": 7,
        "objectType": "Mine",
        "passable": Array [
          255,
          255,
          255,
          255,
          255,
          31,
        ],
      },
      "body": Object {
        "data": Object {
          "owner": 64,
        },
        "type": "ResourceGenerator",
      },
      "oaIndex": 11,
      "x": 23,
      "y": 14,
      "z": 0,
    },
    Object {
      "attributes": Object {
        "above": 0,
        "active": Array [
          0,
          0,
          0,
          0,
          0,
          64,
        ],
        "allowedLandscapes": 511,
        "header": "AVXabnd0.def",
        "landscapeGroup": 1,
        "metaType": "ResourceGenerator",
        "objectGroup": "",
        "objectSubType": 7,
        "objectType": "Mine",
        "passable": Array [
          255,
          255,
          255,
          255,
          255,
          31,
        ],
      },
      "body": Object {
        "data": Object {
          "owner": 126,
        },
        "type": "ResourceGenerator",
      },
      "oaIndex": 11,
      "x": 27,
      "y": 14,
      "z": 0,
    },
    Object {
      "attributes": Object {
        "above": 0,
        "active": Array [
          0,
          0,
          0,
          0,
          0,
          64,
        ],
        "allowedLandscapes": 511,
        "header": "AVMalch0.def",
        "landscapeGroup": 7,
        "metaType": "ResourceGenerator",
        "objectGroup": "",
        "objectSubType": 1,
        "objectType": "Mine",
        "passable": Array [
          255,
          255,
          255,
          255,
          255,
          31,
        ],
      },
      "body": Object {
        "data": Object {
          "owner": 0,
        },
        "type": "ResourceGenerator",
      },
      "oaIndex": 12,
      "x": 3,
      "y": 17,
      "z": 0,
    },
    Object {
      "attributes": Object {
        "above": 0,
        "active": Array [
          0,
          0,
          0,
          0,
          0,
          64,
        ],
        "allowedLandscapes": 511,
        "header": "AVMcrdr0.def",
        "landscapeGroup": 1,
        "metaType": "ResourceGenerator",
        "objectGroup": "",
        "objectSubType": 4,
        "objectType": "Mine",
        "passable": Array [
          255,
          255,
          255,
          255,
          255,
          31,
        ],
      },
      "body": Object {
        "data": Object {
          "owner": 1,
        },
        "type": "ResourceGenerator",
      },
      "oaIndex": 13,
      "x": 6,
      "y": 17,
      "z": 0,
    },
    Object {
      "attributes": Object {
        "above": 1,
        "active": Array [
          0,
          0,
          0,
          0,
          0,
          64,
        ],
        "allowedLandscapes": 511,
        "header": "AVMgedr0.def",
        "landscapeGroup": 1,
        "metaType": "ResourceGenerator",
        "objectGroup": "",
        "objectSubType": 5,
        "objectType": "Mine",
        "passable": Array [
          255,
          255,
          255,
          255,
          31,
          31,
        ],
      },
      "body": Object {
        "data": Object {
          "owner": 2,
        },
        "type": "ResourceGenerator",
      },
      "oaIndex": 14,
      "x": 9,
      "y": 17,
      "z": 0,
    },
    Object {
      "attributes": Object {
        "above": 0,
        "active": Array [
          0,
          0,
          0,
          0,
          0,
          64,
        ],
        "allowedLandscapes": 511,
        "header": "AVMgodr0.def",
        "landscapeGroup": 1,
        "metaType": "ResourceGenerator",
        "objectGroup": "",
        "objectSubType": 6,
        "objectType": "Mine",
        "passable": Array [
          255,
          255,
          255,
          255,
          255,
          31,
        ],
      },
      "body": Object {
        "data": Object {
          "owner": 3,
        },
        "type": "ResourceGenerator",
      },
      "oaIndex": 15,
      "x": 12,
      "y": 17,
      "z": 0,
    },
    Object {
      "attributes": Object {
        "above": 1,
        "active": Array [
          0,
          0,
          0,
          0,
          0,
          64,
        ],
        "allowedLandscapes": 511,
        "header": "AVMordr0.def",
        "landscapeGroup": 1,
        "metaType": "ResourceGenerator",
        "objectGroup": "",
        "objectSubType": 2,
        "objectType": "Mine",
        "passable": Array [
          255,
          255,
          255,
          255,
          31,
          31,
        ],
      },
      "body": Object {
        "data": Object {
          "owner": 4,
        },
        "type": "ResourceGenerator",
      },
      "oaIndex": 16,
      "x": 15,
      "y": 17,
      "z": 0,
    },
    Object {
      "attributes": Object {
        "above": 0,
        "active": Array [
          0,
          0,
          0,
          0,
          0,
          64,
        ],
        "allowedLandscapes": 511,
        "header": "AVMsawd0.def",
        "landscapeGroup": 1,
        "metaType": "ResourceGenerator",
        "objectGroup": "",
        "objectSubType": 0,
        "objectType": "Mine",
        "passable": Array [
          255,
          255,
          255,
          255,
          63,
          15,
        ],
      },
      "body": Object {
        "data": Object {
          "owner": 5,
        },
        "type": "ResourceGenerator",
      },
      "oaIndex": 17,
      "x": 19,
      "y": 17,
      "z": 0,
    },
    Object {
      "attributes": Object {
        "above": 0,
        "active": Array [
          0,
          0,
          0,
          0,
          0,
          64,
        ],
        "allowedLandscapes": 511,
        "header": "AVMsulf0.def",
        "landscapeGroup": 7,
        "metaType": "ResourceGenerator",
        "objectGroup": "",
        "objectSubType": 3,
        "objectType": "Mine",
        "passable": Array [
          255,
          255,
          255,
          255,
          255,
          31,
        ],
      },
      "body": Object {
        "data": Object {
          "owner": 6,
        },
        "type": "ResourceGenerator",
      },
      "oaIndex": 18,
      "x": 22,
      "y": 17,
      "z": 0,
    },
  ],
}
`;
